### DISCUSSION TITLE
Split 'OTA Provider' into seperate code files per manufacturer and utilize zha-device-handlers to maintain and distribute custom 'OTA Providers'?

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#654] [Ideas]
### Posted this question in https://github.com/zigpy/zigpy/issues/154 but perhaps it would be better as a separate discussion here instead:

Would it maybe be a good idea if [zha-device-handlers](https://github.com/zigpy/zha-device-handlers) repo could also host custom 'OTA Provider' (OTAProvider) for Zigbee devices?

Could make it easier to submit a new seperate 'OTA Provider' to [zha-device-handlers](https://github.com/zigpy/zha-device-handlers) or add custom 'OTA Providers' without updating zigpy?

Regardless, might it not also be a good idea to split each 'OTA Provider' into a separate code file per manufacturer for readability?

https://github.com/zigpy/zigpy/tree/dev/zigpy/ota

That is, refactor provider.py to be more modular so that it in turn point to separate Python code files for each OTA provider source?

Please see issue https://github.com/zigpy/zha-device-handlers/issues/750 which raised a related question as user posted a request for a new OTA Provider there.

PS: I believe that some ZHA users today already think that "ZHA Device Handlers" also handle 'OTA Providers' sources and not zigpy.

**Comment by Adminiuga** [comment]
No, just submit a PR to zigpy.

**Comment by mdeweerd** [comment]
I just got a notification about the pull request #994 being merged.

I think that [Koenkk/Zigbee-OTA](/Koenkk/Zigbee-OTA) is a good source for OTA upgrades and you can look at/integrate [zha_toolkit/ota.py](https://github.com/mdeweerd/zha-toolkit/blob/main/custom_components/zha_toolkit/ota.py) that downloads binaries listed there if the firmware is potentially valid for devices that are present (using the manufacturer code).

**Reply by puddly** [reply]
Thanks, I will take a look at it.

I'm going to be completely rewriting OTA in a future zigpy release to be device-centric. The current approach isn't maintainable and will not integrate with ZHA properly.

**Reply by MattWestb** [reply]
Z2M is not verifying that the files is working OK then adding then and some tuya MCU devices (the Zigbee module) and Xiaomi / Aqara devices (i think files was sniffed and patches for not using the right device ID) and  have broken for users.
ZHA is only using official files for automatic OTA updates and the user must self download the OTA files if like using "other" files on own risk.

"device-centric" sound interesting and also possible downgrading devices :-)))

**Reply by mdeweerd** [reply]
@MattWestb I agree taht Z2M is not verifying the files.  It is still a good source to get binaries from and it should be made available to end users.  The best way would be to propose the update to users from the device view and maybe indicate a warning there that the file is from a community source that is not as reliable.

It would even be nice to allow user to vote up or down on a firmware update and show that as an indicator, whether the source is official or not, and show that in the device view.

But in the mean time, users are struggling to download updates and place them in the right location.

Ideally, updating a device would require a voluntary action from the user and not be applied just because the update is available.

Anyway, there are steps, and zigbee-OTA could be added as a provider just as ikea or ledvance.

**Reply by Hedda** [reply]
> I'm going to be completely rewriting OTA in a future zigpy release to be device-centric. The current approach isn't maintainable and will not integrate with ZHA properly.

I assume that ”device centric” approach will mean it will enable a user GUI/UI experince in Home Assistant with ZHA more similar to how Z-Wave OTA updates can work in the latest Z-Wave JS integration there? As discussed here -> https://github.com/zigpy/zigpy/discussions/951

**Reply by Hedda** [reply]
> The best way would be to propose the update to users from the device view and maybe indicate a warning there that the file is from a community source that is not as reliable.

> Ideally, updating a device would require a voluntary action from the user and not be applied just because the update is available.

Home Assistant newish update framework does allow that and the Z-Wave integration in Home Assistant recently implemented that for device centric specific updates, again see discussion -> https://github.com/zigpy/zigpy/discussions/951

> It would even be nice to allow user to vote up or down on a firmware update and show that as an indicator, whether the source is official or not, and show that in the device view.

That would really be next-next-next level, but FYI, Home Assistant newish updates framework at least already allow users to postpone/ignore specific device centric updates until the next major release of Home Assistant (again already implemented for the Z-Wave JS integration).

**Comment by Hedda** [comment]
FYI, be sure to check out puddly's draft PR here which will add support for Generic OTA providers -> https://github.com/zigpy/zigpy/pull/1165

> Implements the format specified in https://github.com/zigpy/zigpy/wiki/OTA-Information-for-Manufacturers:
> 
> ```yaml
> zha:
>   zigpy_config:
>     ota:
>       remote_providers:
>         - url: "https://fw.zigbee.example.org/ota.json"
>           manufacturer_ids: [0x1234, 0x5678]
>         - url: "https://fw.zigbee.example.org/ota-beta.json"
> ```
> 
> This allows for new providers to be added to zigpy without extra code. This also allows for manufacturers to distribute test feed URLs to customers without deploying them globally.

puddly has also added a new feature to the "[zigpy-cli](https://github.com/zigpy/zigpy-cli)" tool that manufacturers can use to produce a skeleton JSON index from a given set of OTA files, and that way make sure manufacturers OTA files are correctly-structured so that they will be compatible with zigpy out-of-the-box. See the README.md for the "[zigpy-cli](https://github.com/zigpy/zigpy-cli)" tool:

https://github.com/zigpy/zigpy-cli/blob/dev/README.md#ota

https://github.com/zigpy/zigpy/wiki/OTA-Information-for-Manufacturers

**Comment by mdeweerd** [comment]
Having a standard format is good.  It does not guarantee that manufacturers will use it, but at least there is the proposal.

As each manufacturer should idealy maintain an OTA files directory, there is still the question of a file listing all those OTA listings so that users do need to go look for them.  There is also the question of how this central listing and the manufacturer listings are "secure".

For instance, "idealy":
- The central listing should indicate which manufacturers are allowed for the referenced manufacturer listing.
- A limited number of trusted people can update the central listing.
- Ideally the Connectivity Standards Alliance (CSA) should really be handling this, but they will be reluctant to do it for devices of manufacturers that are not "authorised" by the CSA, and - also - not all FW is published by their manufacturers (some is reconstructed from sniffs as I understand it) so there would still be a need to maintain another listing somewhere.

**Reply by MattWestb** [reply]
Many of the points is implanted in Matter and some can being implanted in Zigbee R23 if CSA is interested then R32 shall being more or less the same for Zigbee and Matter or perhaps being done in one Matter bridge from the manufacture for there Zigbee devices and other brands can using the same firmware in there matter bridges for updating all known /  certed device firmware.

Regarding of the security if the manufacture is using one good hardware and implanting the OTA in on good way it shall not being any problems. You can encrypting the firmware so no one can doing changes in it or signing it like IKEA have done from day 1. 
Also with more resent hardware you have good strong protected storage for keys that making it nearly impossible braking the firmware change from the  factory to  the flashed in the device.

Then manufactures is not certing there devices (most Xiaomi and tuya devices) or releasing bad firmware that is making devices cat being updated with OTA (IKEA with some controller need being reversed with SWD flashing older firmware) we cant do  so mush (and Xiaomi OTAs is sniffed and is not having manufacturer and image ID set so user is  flashing them on wrong devices => braking devices).

I have sniffing IKEA OTA from Dirigera (no CSA-certed firmware in secured clod storage) that is using different files and we have putting them together and the signing is still OK so the device is flashing it then have downloading it OK. If the signing was broken then the device is throwing the firmware and waiting  for new request = no security problems as long the signing key is leaked / hacked (its still in the device flash / firmware).

In the end we cant changing how manufacture have doing there systems and we musts living with it but some can get inspired and changing there way of doing things (i hope Philips is releasing there OTA to the public  feed and we dont need hacking there ZHGW).
